synapse/rest/client/v1/login.py

```diff
@@ -144,99 +141,49 @@ async def _do_other_login(self, login_submission):
         # Convert deprecated authdict formats to the current scheme
         client_dict_convert_legacy_fields_to_identifier(login_submission)
 
-        if "identifier" not in login_submission:
-            raise SynapseError(400, "Missing param: identifier")
-
-        identifier = login_submission["identifier"]
-        if "type" not in identifier:
-            raise SynapseError(400, "Login identifier has no type")
-
-        # convert phone type identifiers to generic threepids
-        if identifier["type"] == "m.id.phone":
-            identifier = login_id_phone_to_thirdparty(identifier)
-
-        # convert threepid identifiers to user IDs
-        if identifier["type"] == "m.id.thirdparty":
-            address = identifier.get("address")
-            medium = identifier.get("medium")
-
-            if medium is None or address is None:
-                raise SynapseError(400, "Invalid thirdparty identifier")
-
-            if medium == "email":
-                # For emails, transform the address to lowercase.
-                # We store all email addreses as lowercase in the DB.
-                # (See add_threepid in synapse/handlers/auth.py)
-                address = address.lower()
-
-            # We also apply account rate limiting using the 3PID as a key, as
-            # otherwise using 3PID bypasses the ratelimiting based on user ID.
-            self._failed_attempts_ratelimiter.ratelimit((medium, address), update=False)
-
-            # Check for login providers that support 3pid login types
-            (
-                canonical_user_id,
-                callback_3pid,
-            ) = await self.auth_handler.check_password_provider_3pid(
-                medium, address, login_submission["password"]
+        # Check whether this attempt uses a threepid, if so, check if our failed attempt
+        # ratelimiter allows another attempt at this time
+        medium, address = (
+            login_submission.get("medium"),
+            login_submission.get("address"),
+        )
+        if medium and address:
+            self._failed_attempts_ratelimiter.ratelimit(
+                (medium.lower(), address.lower()), update=False
             )
-            if canonical_user_id:
-                # Authentication through password provider and 3pid succeeded
 
-                result = await self._complete_login(
-                    canonical_user_id, login_submission, callback_3pid
-                )
-                return result
```

Comment:
This code path seems to go through an additional `self.auth_handler.validate_login`, I don't know if that's OK or not.

Reply:
Aha, yes good spot.

I believe this originally did not as we have already validated the password via the password auth provider, so there's no need to `validate_login` at that point (which will itself ask password auth providers with a user_id/password combo).

Considering this would change Synapse's behaviour related to third-party code we should probably try to get the behaviour to match.